<font color=Red>**Note**</font>: In order to make your answer stand out, please use colored text to write your answer. For example, the following code sets font color to blue.

<font color=Blue>我是蓝色</font>

**You should put answer to every question in a different cell, so that the result can be displayed clearly. You can also add cells when necessary.** 

**Rename your Jupyter Notebook file as classNumber_Name, and send it to this email address(1614688593@qq.com) before due date.**

<font color=Red>**Due Date: 2021-03-31 23:59**</font>

# Lab 1 Dataset manipulation with Pandas

This tutorial contains 34 questions. All need to be answered except for Question 18, which is optional.

## 1. Load dataset and overview
Load the video game sales dataset from the vgsales.csv file into a DataFrame object.

**【Q1】Call the common attributes of DataFrame and check their outputs (df.index, df.columns, df.values, df.dtypes, df.ndim, df.shape, df.size).**

**【Q2】How many records and how many attributes does it have?**

In [91]:
import pandas as pd

df=pd.read_csv('./vgsales.csv')
print('df.index:\n',df.index)
print('df.columns:\n',df.columns)
print('df.values:\n',df.values)
print('df.dtypes:\n',df.dtypes)
print('df.ndim:\n',df.ndim)
print('df.shape:\n',df.shape)
print('df.size:\n',df.size)

df.index:
 RangeIndex(start=0, stop=16598, step=1)
df.columns:
 Index(['Rank', 'Name', 'Platform', 'Year', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales'],
      dtype='object')
df.values:
 [[1 'Wii Sports' 'Wii' ... 3.77 8.46 82.74]
 [2 'Super Mario Bros.' 'NES' ... 6.81 0.77 40.24]
 [3 'Mario Kart Wii' 'Wii' ... 3.79 3.31 35.82]
 ...
 [16598 'SCORE International Baja 1000: The Official Game' 'PS2' ... 0.0
  0.0 0.01]
 [16599 'Know How 2' 'DS' ... 0.0 0.0 0.01]
 [16600 'Spirits & Spells' 'GBA' ... 0.0 0.0 0.01]]
df.dtypes:
 Rank              int64
Name             object
Platform         object
Year            float64
Genre            object
Publisher        object
NA_Sales        float64
EU_Sales        float64
JP_Sales        float64
Other_Sales     float64
Global_Sales    float64
dtype: object
df.ndim:
 2
df.shape:
 (16598, 11)
df.size:
 182578


In [7]:
print('records:',df.shape[0])
print('attributes:',df.shape[1])

records: 16598
attributes: 11


Describe the dataset with the df.describe() method. This method computes simple statistics of the numeric attributes. 

**【Q3】Does the output includes every attribute? If not, which ones are missing, why? What did you find from the simple statistics?**

In [8]:
df.describe()

Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16598.000000  16327.000000  16598.000000  16598.000000  16598.000000   
mean    8300.605254   2006.406443      0.264667      0.146652      0.077782   
std     4791.853933      5.828981      0.816683      0.505351      0.309291   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4151.250000   2003.000000      0.000000      0.000000      0.000000   
50%     8300.500000   2007.000000      0.080000      0.020000      0.000000   
75%    12449.750000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2020.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16598.000000  16598.000000  
mean       0.048063      0.537441  
std        0.188588      1.555028  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.470000  
max       10.570000     82.740000

<font color=Blue>
No, the output just include numeric attributes. 'Name','Platform','Genre','Publisher' are missing, because the dtypes of them is object rather than float64. I find that the total video games is 16598, the max Global_Sales is 82.74 and the min of Global_Sales is 0.01, the records of the table is from 1980 to 2020.
</font>

**Type of data**
1. numeric/continuous
2. nominal/categorical/discrate(a finite pre-specified set of data)



## 2. Data selection

### 2.1 Select one or more rows
- **【Q4】print video games with the top five annual sales (the first 5 rows)**
- **【Q5】print video games with the 5-10 sales (records 5-10)**
- **【Q6】print video games with the least five annual sales (the last 5 rows)**
- **【Q7】print video games with the 1, 10, 50, and 100 sales (index 0, 9, 49 and 99 respectively)**

You can try with different methods, and compare their usage.

In [10]:
print('Q4:\n',df.head(5))
print('Q5:\n',df[5:10])
print('Q6:\n',df.tail(5))
print('Q7:\n',df.iloc[0],df.iloc[9],df.iloc[49],df.loc[99])

Q4:
    Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37  
Q5:
    Rank                       Name Platform    Year     Genre Publisher  \
5     6                     Tetris       GB  1989.0    Puzzle  Nintendo   
6    

### 2.2 Select one or more columns
- **【Q8】print video games's names (i.e. the Name column)**
- **【Q9】print the Name, Year and Global_Sales columns**

In [14]:
df['Name']
# df.Name

0                                              Wii Sports
1                                       Super Mario Bros.
2                                          Mario Kart Wii
3                                       Wii Sports Resort
4                                Pokemon Red/Pokemon Blue
                               ...                       
16593                  Woody Woodpecker in Crazy Castle 5
16594                       Men in Black II: Alien Escape
16595    SCORE International Baja 1000: The Official Game
16596                                          Know How 2
16597                                    Spirits & Spells
Name: Name, Length: 16598, dtype: object

In [15]:
df[['Name','Year','Global_Sales']]

Name    Year  Global_Sales
0                                            Wii Sports  2006.0         82.74
1                                     Super Mario Bros.  1985.0         40.24
2                                        Mario Kart Wii  2008.0         35.82
3                                     Wii Sports Resort  2009.0         33.00
4                              Pokemon Red/Pokemon Blue  1996.0         31.37
...                                                 ...     ...           ...
16593                Woody Woodpecker in Crazy Castle 5  2002.0          0.01
16594                     Men in Black II: Alien Escape  2003.0          0.01
16595  SCORE International Baja 1000: The Official Game  2008.0          0.01
16596                                        Know How 2  2010.0          0.01
16597                                  Spirits & Spells  2003.0          0.01

[16598 rows x 3 columns]

### 2.3 Select chunks
- **【Q10】print the 2-5 attributes of records 2-4**

Note: if you want to use loc(), you need to figure out the names of columns 2-4.  

In [17]:
df.iloc[2:4,[2,3,4,5]]

Platform    Year   Genre Publisher
2      Wii  2008.0  Racing  Nintendo
3      Wii  2009.0  Sports  Nintendo

### 2.4 Select with conditional statements

- **【Q11】Print sales records for 'Super Mario Bros.'**
- **【Q12】Print action games published in 2006.**
- **【Q13】Print the names of the action games published in 2006.**
- **【Q14】Which games were published before 2000 and obtained a sales record greater than 30 million?**
- **【Q15】Which action games obtained a sales record greater than 20 million?**

In [20]:
df[df.Name=='Super Mario Bros.']

Rank               Name Platform    Year     Genre Publisher  NA_Sales  \
1       2  Super Mario Bros.      NES  1985.0  Platform  Nintendo     29.08   
200   201  Super Mario Bros.       GB  1999.0  Platform  Nintendo      3.40   

     EU_Sales  JP_Sales  Other_Sales  Global_Sales  
1        3.58      6.81         0.77         40.24  
200      1.30      0.15         0.22          5.07

In [21]:
df.loc[(df.Genre=='Action') & (df.Year==2006),:]

Rank                                               Name Platform  \
100      101             The Legend of Zelda: Twilight Princess      Wii   
198      199                Grand Theft Auto: Vice City Stories      PSP   
360      361             Grand Theft Auto: Liberty City Stories      PS2   
556      557            LEGO Star Wars II: The Original Trilogy      PS2   
569      570                                     Hannah Montana       DS   
...      ...                                                ...      ...   
15997  16000                                          Black Cat      PS2   
16023  16026  Rengoku II: The Stairway to H.E.A.V.E.N. (jp s...      PSP   
16035  16038                                    Kaitou Rousseau       DS   
16230  16233           Hi Hi Puffy AmiYumi: The Genie & the Amp       DS   
16413  16416                           The Godfather (JP sales)     X360   

         Year   Genre                     Publisher  NA_Sales  EU_Sales  \
100    2006.0  Action                      Nintendo      3.83      2.19   
198    2006.0  Action          Take-Two Interactive      1.70      2.02   
360    2006.0  Action          Take-Two Interactive      1.56      1.40   
556    2006.0  Action                     LucasArts      1.85      0.64   
569    2006.0  Action    Disney Interactive Studios      1.59      0.80   
...       ...     ...                           ...       ...       ...   
15997  2006.0  Action                        Capcom      0.00      0.00   
16023  2006.0  Action  Konami Digital Entertainment      0.00      0.00   
16035  2006.0  Action            Namco Bandai Games      0.00      0.00   
16230  2006.0  Action                   D3Publisher      0.01      0.00   
16413  2006.0  Action               Electronic Arts      0.00      0.00   

       JP_Sales  Other_Sales  Global_Sales  
100        0.60         0.70          7.31  
198        0.16         1.21          5.08  
360        0.07         0.50          3.54  
556        0.01         0.19          2.69  
569        0.00         0.26          2.65  
...         ...          ...           ...  
15997      0.01         0.00          0.01  
16023      0.01         0.00          0.01  
16035      0.01         0.00          0.01  
16230      0.00         0.00          0.01  
16413      0.01         0.00          0.01  

[184 rows x 11 columns]

In [22]:
df.Name[df.Year==2006]

0                                    Wii Sports
6                         New Super Mario Bros.
7                                      Wii Play
20                Pokemon Diamond/Pokemon Pearl
100      The Legend of Zelda: Twilight Princess
                          ...                  
16513                           Deal or No Deal
16520      Tengai Makyo: Dai Yon no Mokushiroku
16523                   Style Book: Cinnamoroll
16547                           Bookworm Deluxe
16568                               XI Coliseum
Name: Name, Length: 1008, dtype: object

In [23]:
df.loc[(df.Year<2000) & (df.Global_Sales>30),:]

Rank                      Name Platform    Year         Genre Publisher  \
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   
5     6                    Tetris       GB  1989.0        Puzzle  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
1     29.08      3.58      6.81         0.77         40.24  
4     11.27      8.89     10.22         1.00         31.37  
5     23.20      2.26      4.22         0.58         30.26

In [24]:
df.loc[(df.Genre=='Action') & (df.Global_Sales>20),:]

Rank                           Name Platform    Year   Genre  \
16    17             Grand Theft Auto V      PS3  2013.0  Action   
17    18  Grand Theft Auto: San Andreas      PS2  2004.0  Action   

               Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
16  Take-Two Interactive      7.01      9.27      0.97         4.14   
17  Take-Two Interactive      9.43      0.40      0.41        10.57   

    Global_Sales  
16         21.40  
17         20.81

## 3. Data manipulation

In this section we will practice add/delete rows and columns from a dataset.

**【Q16】Create a data subset (vg50) by selecting the first 50 rows and the Name, Year, Publisher columns, i.e. the basic information of a game.**

In [64]:
vg50=df.loc[:49,['Name','Year','Publisher']]

### 3.1 Add records
- **【Q17】Add the last 10 records from the original vg dataset to the vg50 subset. Print the last 15 records of vg50. What did you find?**

In [65]:
vg50=vg50.append(df.loc[:,['Name','Year','Publisher']].tail(10),ignore_index=True)
vg50
# pd.concat([vg50,vg.tail(10)],sort=True,ignore_index=True)

Name    Year  \
0                                          Wii Sports  2006.0   
1                                   Super Mario Bros.  1985.0   
2                                      Mario Kart Wii  2008.0   
3                                   Wii Sports Resort  2009.0   
4                            Pokemon Red/Pokemon Blue  1996.0   
5                                              Tetris  1989.0   
6                               New Super Mario Bros.  2006.0   
7                                            Wii Play  2006.0   
8                           New Super Mario Bros. Wii  2009.0   
9                                           Duck Hunt  1984.0   
10                                         Nintendogs  2005.0   
11                                      Mario Kart DS  2005.0   
12                        Pokemon Gold/Pokemon Silver  1999.0   
13                                            Wii Fit  2007.0   
14                                       Wii Fit Plus  2009.0   
15                                 Kinect Adventures!  2010.0   
16                                 Grand Theft Auto V  2013.0   
17                      Grand Theft Auto: San Andreas  2004.0   
18                                  Super Mario World  1990.0   
19       Brain Age: Train Your Brain in Minutes a Day  2005.0   
20                      Pokemon Diamond/Pokemon Pearl  2006.0   
21                                   Super Mario Land  1989.0   
22                                Super Mario Bros. 3  1988.0   
23                                 Grand Theft Auto V  2013.0   
24                        Grand Theft Auto: Vice City  2002.0   
25                      Pokemon Ruby/Pokemon Sapphire  2002.0   
26                        Pokemon Black/Pokemon White  2010.0   
27        Brain Age 2: More Training in Minutes a Day  2005.0   
28                             Gran Turismo 3: A-Spec  2001.0   
29                     Call of Duty: Modern Warfare 3  2011.0   
30            Pokémon Yellow: Special Pikachu Edition  1998.0   
31                            Call of Duty: Black Ops  2010.0   
32                                Pokemon X/Pokemon Y  2013.0   
33                          Call of Duty: Black Ops 3  2015.0   
34                         Call of Duty: Black Ops II  2012.0   
35                         Call of Duty: Black Ops II  2012.0   
36                     Call of Duty: Modern Warfare 2  2009.0   
37                     Call of Duty: Modern Warfare 3  2011.0   
38                               Grand Theft Auto III  2001.0   
39                            Super Smash Bros. Brawl  2008.0   
40                            Call of Duty: Black Ops  2010.0   
41                        Animal Crossing: Wild World  2005.0   
42                                       Mario Kart 7  2011.0   
43                                             Halo 3  2007.0   
44                                 Grand Theft Auto V  2014.0   
45               Pokemon HeartGold/Pokemon SoulSilver  2009.0   
46                                     Super Mario 64  1996.0   
47                                     Gran Turismo 4  2004.0   
48                                 Super Mario Galaxy  2007.0   
49          Pokemon Omega Ruby/Pokemon Alpha Sapphire  2014.0   
50                                   Mega Brain Boost  2008.0   
51  Chou Ezaru wa Akai Hana: Koi wa Tsuki ni Shiru...  2016.0   
52  Eiyuu Densetsu: Sora no Kiseki Material Collec...  2007.0   
53                                Myst IV: Revelation  2004.0   
54                                           Plushees  2008.0   
55                 Woody Woodpecker in Crazy Castle 5  2002.0   
56                      Men in Black II: Alien Escape  2003.0   
57   SCORE International Baja 1000: The Official Game  2008.0   
58                                         Know How 2  2010.0   
59                                   Spirits & Spells  2003.0   

                      Publisher  
0                      Nintendo  
1              

<font color=Blue>
I find that:
1. The most best seller is 'Wii Sports' and the worst seller is 'Spirts & Spells';
2. The records sort by 'Global_Sales' in descesng order.
</font>

If we would like continuous index values, we can either set the ignore_index attribute to True when calling the append() method, or use the reset_index() method to reset the index values. Try one of the options and print the last 15 records of the vg50 dataset.

In [66]:
vg50.reset_index().tail(15)

index                                               Name    Year  \
45     45               Pokemon HeartGold/Pokemon SoulSilver  2009.0   
46     46                                     Super Mario 64  1996.0   
47     47                                     Gran Turismo 4  2004.0   
48     48                                 Super Mario Galaxy  2007.0   
49     49          Pokemon Omega Ruby/Pokemon Alpha Sapphire  2014.0   
50     50                                   Mega Brain Boost  2008.0   
51     51  Chou Ezaru wa Akai Hana: Koi wa Tsuki ni Shiru...  2016.0   
52     52  Eiyuu Densetsu: Sora no Kiseki Material Collec...  2007.0   
53     53                                Myst IV: Revelation  2004.0   
54     54                                           Plushees  2008.0   
55     55                 Woody Woodpecker in Crazy Castle 5  2002.0   
56     56                      Men in Black II: Alien Escape  2003.0   
57     57   SCORE International Baja 1000: The Official Game  2008.0   
58     58                                         Know How 2  2010.0   
59     59                                   Spirits & Spells  2003.0   

                      Publisher  
45                     Nintendo  
46                     Nintendo  
47  Sony Computer Entertainment  
48                     Nintendo  
49                     Nintendo  
50        Majesco Entertainment  
51              dramatic create  
52           Falcom Corporation  
53                      Ubisoft  
54                    Destineer  
55                        Kemco  
56                   Infogrames  
57                   Activision  
58                     7G//AMES  
59                      Wanadoo

### 3.2 Add columns

- **【Q18】Add the corresponding global sales from the original dataset to the vg50 subset, and set the new column name to Sales.** 

In [68]:
vg50['Sales']=pd.concat([df['Global_Sales'][:50],df['Global_Sales'].tail(10)]).reset_index(drop=True)
vg50.tail(12)
# pd.concat([df['Global_Sales'][:50],df['Global_Sales'].tail(10)]).reset_index(drop=True)

Name    Year  \
48                                 Super Mario Galaxy  2007.0   
49          Pokemon Omega Ruby/Pokemon Alpha Sapphire  2014.0   
50                                   Mega Brain Boost  2008.0   
51  Chou Ezaru wa Akai Hana: Koi wa Tsuki ni Shiru...  2016.0   
52  Eiyuu Densetsu: Sora no Kiseki Material Collec...  2007.0   
53                                Myst IV: Revelation  2004.0   
54                                           Plushees  2008.0   
55                 Woody Woodpecker in Crazy Castle 5  2002.0   
56                      Men in Black II: Alien Escape  2003.0   
57   SCORE International Baja 1000: The Official Game  2008.0   
58                                         Know How 2  2010.0   
59                                   Spirits & Spells  2003.0   

                Publisher  Sales  
48               Nintendo  11.52  
49               Nintendo  11.33  
50  Majesco Entertainment   0.01  
51        dramatic create   0.01  
52     Falcom Corporation   0.01  
53                Ubisoft   0.01  
54              Destineer   0.01  
55                  Kemco   0.01  
56             Infogrames   0.01  
57             Activision   0.01  
58               7G//AMES   0.01  
59                Wanadoo   0.01

### 3.3 Update values

- **【Q19】Set the first record's year to 2004.**
- **【Q20】Change the sales unit from million dollars to dollars (i.e. times 1,000,000)**

In [70]:
df.loc[0,'Year']=2004
df.head(1)

Rank        Name Platform    Year   Genre Publisher  NA_Sales  EU_Sales  \
0     1  Wii Sports      Wii  2004.0  Sports  Nintendo     41.49     29.02   

   JP_Sales  Other_Sales  Global_Sales  
0      3.77         8.46         82.74

In [78]:
def fun(ele):
    return ele*1000000.

df['Global_Sales'].map(fun).head(5)

0    82740000.0
1    40240000.0
2    35820000.0
3    33000000.0
4    31370000.0
Name: Global_Sales, dtype: float64

### 3.4 Compute from existing values

Firstly, **create a new subset with sales records in 2009 (1431 in total)**. For demonstration purpose, here we only select the regional sales columns, i.e. excluding the Global_Sales column. We are going to compute this column.

In [79]:
subset=df.loc[df.Year==2009,['NA_Sales','EU_Sales','JP_Sales','Other_Sales']]
subset

NA_Sales  EU_Sales  JP_Sales  Other_Sales
3         15.75     11.01      3.28         2.96
8         14.59      7.06      4.70         2.26
14         9.09      8.59      2.53         1.79
36         8.52      3.63      0.08         1.29
45         4.40      2.77      3.96         0.77
...         ...       ...       ...          ...
16540      0.00      0.01      0.00         0.00
16551      0.00      0.00      0.01         0.00
16577      0.00      0.01      0.00         0.00
16586      0.00      0.01      0.00         0.00
16587      0.00      0.00      0.01         0.00

[1431 rows x 4 columns]

- **【Q21】compute the total sales of each game, i.e. the Global_Sales column's value in the original dataset. Compare the computed value with the original value in the dataset.**
- **【Q22】compute the total sales for each region in 2009**

In [80]:
subset['Global_Sales']=subset.apply(lambda x:x.sum(),axis=1)
subset.head(10)

NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales
3       15.75     11.01      3.28         2.96         33.00
8       14.59      7.06      4.70         2.26         28.61
14       9.09      8.59      2.53         1.79         22.00
36       8.52      3.63      0.08         1.29         13.52
45       4.40      2.77      3.96         0.77         11.90
55       4.99      3.69      0.38         1.63         10.69
83       0.98      6.42      0.00         0.71          8.11
102      3.51      3.03      0.00         0.73          7.27
119      3.27      2.25      0.21         1.00          6.73
133      4.34      1.35      0.06         0.61          6.36

In [84]:
subset['Global_Sales'].sum()

667.1

### 3.5 Drop columns and rows

- **【Q23】drop the last column**
- **【Q24】drop columns whose name starts with S**
- **【Q25】drop the 50th record**
- **【Q26】drop records in 2009**

Note: Before drop the columns/records, it is better to check out what are they and how many.

In [92]:
df.drop(df.columns[df.shape[1]-1],axis=1)

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  
0             8.46  
1             0.77  
2             3.31  
3             2.96  
4             1.00  
...            ...  
16593         0.00  
16594         0.00  
16595         0.00  
16596         0.00  
16597         0.00  

[16598 rows x 10 columns]

In [96]:
import numpy as np
df['Sport']=np.zeros(df.shape[0])
df.drop([c for c in df.columns if c[:1]=='S'],axis=1)
# df = df[df.columns.drop(list(df.filter(regex='Test\d+')))]
# vg50.loc[:,~vg50.columns.str.startwith('s')]

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

In [97]:
df.drop([50])

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  Sport  
0             8.46         82.74    0.0  
1             0.77         40.24    0.0  
2             3.31         35.82    0.0  
3             2.96         33.00    0.0  
4             1.00         31.37    0.0  
...            ...           ...    ...  
16593         0.00          0.01    0.0  
16594         0.00          0.01    0.0  
16595         0.00          0.01    0.0  
16596         0.00          0.01    0.0  
16597         0.00          0.01    0.0  

[16597 rows x 12 columns]

In [98]:
df.drop(df[df.Year==2009].index)

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
5          6                                            Tetris       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
5      1989.0        Puzzle    Nintendo     23.20      2.26      4.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  Sport  
0             8.46         82.74    0.0  
1             0.77         40.24    0.0  
2             3.31         35.82    0.0  
4             1.00         31.37    0.0  
5             0.58         30.26    0.0  
...            ...           ...    ...  
16593         0.00          0.01    0.0  
16594         0.00          0.01    0.0  
16595         0.00          0.01    0.0  
16596         0.00          0.01    0.0  
16597         0.00          0.01    0.0  

[15167 rows x 12 columns]

## 4. Sort data

- **【Q27】What are the top five games sold in Europe in 2009?**
- **【Q28】What are the top five games sold in North America in 2009?**
- **【Q29】What are the top five games sold in Japan in 2009?**

In [99]:
df[df.Year==2009].sort_values('EU_Sales',ascending=False).head(5)

Rank                            Name Platform    Year       Genre  \
3      4               Wii Sports Resort      Wii  2009.0      Sports   
14    15                    Wii Fit Plus      Wii  2009.0      Sports   
8      9       New Super Mario Bros. Wii      Wii  2009.0    Platform   
83    84                      The Sims 3       PC  2009.0  Simulation   
55    56  Call of Duty: Modern Warfare 2      PS3  2009.0     Shooter   

          Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
3          Nintendo     15.75     11.01      3.28         2.96         33.00   
14         Nintendo      9.09      8.59      2.53         1.79         22.00   
8          Nintendo     14.59      7.06      4.70         2.26         28.62   
83  Electronic Arts      0.98      6.42      0.00         0.71          8.11   
55       Activision      4.99      3.69      0.38         1.63         10.69   

    Sport  
3     0.0  
14    0.0  
8     0.0  
83    0.0  
55    0.0

In [ ]:
df[df.Year==2009].sort_values('NA_Sales',ascending=False).head(5)

In [ ]:
df[df.Year==2009].sort_values('JP_Sales',ascending=False).head(5)

## 5. Data grouping

Use the original video game sales dataset, investigate the following questions.

- **【Q30】What are the sales for each game genre? What are the top three popular game genres globally and across the years?**

Here we only care about the genre and global sales columns, so we can firstly filter these columns. Otherwise, all columns will be aggreated, e.g. sumed, which is unnecessary.

In [105]:
vg=pd.read_csv('./vgsales.csv')
vg.loc[:,['Genre','Global_Sales']].groupby('Genre').agg({'Global_Sales':'sum'}).sort_values('Global_Sales',ascending=False).head(3)

Global_Sales
Genre                
Action        1751.18
Sports        1330.93
Shooter       1037.37

- **【Q31】What are the annual sales for different publishers? What are the largest three game publishers?**

Now we only care about the publishers and the sales columns, so like above, we can firstly filter these columns. 

In [106]:
vg[['Publisher','Global_Sales']].groupby('Publisher').agg({'Global_Sales':'sum'}).sort_values('Global_Sales',ascending=False).head(3)

Global_Sales
Publisher                    
Nintendo              1786.56
Electronic Arts       1110.32
Activision             727.46

- **【Q32】What are the total sales for each year?**

Likewise, we first create a subset.

In [108]:
vg[['Year','Global_Sales']].groupby('Year').agg({'Global_Sales':'sum'})

Global_Sales
Year                
1980.0         11.38
1981.0         35.77
1982.0         28.86
1983.0         16.79
1984.0         50.36
1985.0         53.94
1986.0         37.07
1987.0         21.74
1988.0         47.22
1989.0         73.45
1990.0         49.39
1991.0         32.23
1992.0         76.16
1993.0         45.98
1994.0         79.17
1995.0         88.11
1996.0        199.15
1997.0        200.98
1998.0        256.47
1999.0        251.27
2000.0        201.56
2001.0        331.47
2002.0        395.52
2003.0        357.85
2004.0        419.31
2005.0        459.94
2006.0        521.04
2007.0        611.13
2008.0        678.90
2009.0        667.30
2010.0        600.45
2011.0        515.99
2012.0        363.54
2013.0        368.11
2014.0        337.05
2015.0        264.44
2016.0         70.93
2017.0          0.05
2020.0          0.29

- **【Q33】What are the games that have been sold for the longest time and what are their global sales like?**

In [110]:
vg[['Name','Year','Global_Sales']].groupby('Name').agg({'Year':'count','Global_Sales':'sum'}).sort_values('Year',ascending=False)

Year  Global_Sales
Name                                                       
Need for Speed: Most Wanted                12         14.08
FIFA 14                                     9         17.14
Ratatouille                                 9          4.46
LEGO Marvel Super Heroes                    9          9.86
Monopoly                                    8          7.70
...                                       ...           ...
The Daring Game for Girls                   0          0.22
Saint                                       0          0.02
TERA                                        0          0.41
Housekeeping                                0          0.04
Shonen Jump's Yu-Gi-Oh! GX Card Almanac     0          0.02

[11493 rows x 2 columns]

## 5. FileIO with Pandas

- **【Q34】Generate a dataset that only contains the game name and its total (i.e. aggregated) sales (that is, the NA_Sales, EU_Sales, JP_Sales, Other_Sales and Global_Sales). Output it to a CSV and to an Excel file respectively. Take screenshots of the output files (e.g. opened in TextEditor and Excel) and paste them below.**

Here we ignore the Rank, Platform, Year, Genre, Publisher columns, and combine sales across different years and platforms for games with the same name.

In [113]:
subset2=vg[['Name','NA_Sales','EU_Sales','JP_Sales','Other_Sales','Global_Sales']].groupby('Name').agg({'NA_Sales':'sum','EU_Sales':'sum','JP_Sales':'sum','Other_Sales':'sum','Global_Sales':'sum'})
subset2.to_csv('subset2.csv')
!pip install xlwt
subset2.to_excel('subset2.xls')

Looking in indexes: http://mirrors.aliyun.com/pypi/simple/
ipykernel_launcher:4: FutureWarning: As the xlwt package is no longer maintained, the xlwt engine will be removed in a future version of pandas. This is the only engine in pandas that supports writing in the xls format. Install openpyxl and write to an xlsx file instead. You can set the option io.excel.xls.writer to 'xlwt' to silence this warning. While this option is deprecated and will also raise a warning, it can be globally set and the warning suppressed.


### Below is screenshot of 'subset2.csv' file
![](./screenshot_of_csv.png)
### Below is screenshot of 'subset2.xls' file
![](./screenshot_of_excel.png)

## 6. Extended manipulation

There are many other possibilities that are not included in our lectures and this tutorial. For example, you can try to integrate multiple datasets using the pd.concat() method. If you have explored other manipulation methods, feel free to write your work below.

In [116]:
pd.__version__

'1.2.4'

In [118]:
data = {1: "You", 2: "And", 3: "Me"}
pd.Series(data, index = [1, 2], name="WhereAmI")

1    You
2    And
Name: WhereAmI, dtype: object

In [119]:
print(vg.to_string())

                                                           Blazer Drive       DS  2008.0  Role-Playing                                    Sega      0.00      0.00      0.01         0.00          0.01
16526  16529                                                                                                                    Onigokko! Portable      PSP  2013.0     Adventure                               Alchemist      0.00      0.00      0.01         0.00          0.01
16527  16530                                                                                                             Hoshizora no Comic Garden       DS  2008.0     Adventure                             D3Publisher      0.00      0.00      0.01         0.00          0.01
16528  16531                                                                                                                     Kiniro no Corda 3      3DS  2015.0     Adventure                              Tecmo Koei      0.00      0.00      0.01   